Add NUCLEO-F031K6 board and example

genbattle opened the pull request:
Hi, I added support for the Nucleo F031k6 board to allow me to develop for it using XPCC. So far I've developed the platform definition based on the existing definition for the f103rb device (including support for other devices in the f031xx family). I've got the platform working with the basic blink example.

Still to be done:

- [x] OpenOCD support; there's no file for the nucleo f031k6 included with OpenOCD currently, so I have developed my own configuration file for this platform that seems to work, but I'm not sure where it fits in within XPCC.
- [x] Tests and examples for this platform.
- [x] Verify serial debugging works properly.

I would appreciate any guidance you can give me on whether this work will be valuable to the project, and how it might be improved (especially with respect to adding OpenOCD support for the platform) before it is merged.

daniel-k commented:
Hi! Did you try the [upstream Nucleo F0 OpenOCD config](http://repo.or.cz/openocd.git/blob/HEAD:/tcl/board/st_nucleo_f0.cfg)?

I'm not sure how @salkinium is handling OpenOCD configs nowadays, but using the upstream configs should be the best idea. Using them is just a matter of:

```bash
$ openocd -f board/st_nucleo_f0.cfg
```

genbattle commented:
I've just added it to the pull request - I missed it when I looking for a configuration that matched the board.

genbattle commented:
Managed to get the virtual serial port working at 115200, so debugging/logging definitely works. Strangely reading the serial port from the command line on linux using `cat` only works for 9600 baud, not 115200. `minicom` works for both though.

daniel-k commented:
> Strangely reading the serial port from the command line on linux using cat only works for 9600 baud

Doesn't surprise me too much. Did you change the baudrate? I guess you can do that with tools such as `stty`, but I've never used that myself actually. That's why we have tools like minicom, picocom and the like I guess.

genbattle commented:
I didn't change the baudrate from the default. I've mostly worked with Windows when it comes to microcontrollers in the past, so I'm still learning the particularities of serial devices under Linux.

salkinium commented:
> whether this work will be valuable to the project

Yes, very much! I'm impressed by your contribution considering how under-documented most of this is ;-P

We are currently working on the next generation of this framework (over at [modm-io](https://github.com/modm-io), it's very much WIP). In the meantime we're forward porting all changes from xpcc to modm, which is licensed under MPLv2:

> We are relicensing all source code from 3-clause BSD in xpcc to MPLv2 in modm.
> https://www.mozilla.org/en-US/MPL/2.0/  
> https://en.wikipedia.org/wiki/Mozilla_Public_License
>
> The MPLv2 license is more restrictive than the BSD license, but less restrictive than (L)GPL:
>
> - it requires copyleft at file level, but derivatives can be released under other licenses (that are compatible with MPL).
> - in contrast to GPL you only have to provide open-source access to the derived code covered under MPL.
> - It contains a patent retaliation clause, similar to that of the Apache license.
>
> We think MPLv2 is a better choice for modm, as it makes it easier for us to keep modm open-source (copyleft), but does _not_ deter programmers, projects or companies to use modm and contribute back like GPL does.

Since this is a non-trivial change to xpcc, **would you be ok with licensing your contributions as MPLv2 as well?**

> So far I've developed the platform definition based on the existing definition for the f103rb device

Hm, about that: Sorry for making you do this manually! The xpcc device file generator (DFG) [outputs this file for the F031|51 devices](https://gist.github.com/salkinium/88c13fbf159028b095339475df2c056e). I didn't add all device files, I'm doing that now.

(The DFG output has been massively [improved for modm](https://github.com/modm-io/modm-devices/tree/wip/niklas/devices/stm32), but it's not backwards compatible).

> I would appreciate any guidance you can give me

Yes, of course:

I ported another Nucleo-32 board in the past ([Nucleo-f303k8](https://github.com/roboterclubaachen/xpcc/blob/develop/src/xpcc/architecture/platform/board/nucleo_f303k8/nucleo_f303k8.hpp)) which should be similar to the F031K6 one. See the [Nucleo-32 User Manual Section 6.11 (page 21)](http://www.st.com/resource/en/user_manual/dm00231744.pdf) for the subtle differences in pinouts.

The F031K6 is very interesting due to its limited resources and a good opportunity to tweak xpcc for low memory consumption. Those 4kB of RAM will likely require some tweaking to the size of the uart buffer in the board config file (256B should be enough, it's blocking anyways) and main stack size, which is [set to ~3kB by default](https://github.com/roboterclubaachen/xpcc/blob/develop/src/xpcc/architecture/platform/driver/core/cortex/driver.xml#L10) (try with `core.cortex.0.main_stack_size = 992` (1024B-32B)). Disabling the `core.cortex.0.enable_hardfault_handler_log` will also reduce hardfault stack size to 32B for the LED blinking and `core.cortex.0.vector_table_in_ram = false` will keep the vector table out of RAM. ([background](https://github.com/modm-io/modm/blob/wip/niklas/src/modm/platform/core/cortex/option/main_stack_size.md)).
Setting `core.cortex.0.allocator` to `block_allocator` will replace the default newlib allocator (dlmalloc) with a much simpler allocator with less overhead ([background](https://github.com/modm-io/modm/blob/wip/niklas/src/modm/platform/core/cortex/option/allocator.md)).

Keep in mind that depending on how much of newlib you use, a few kB of static memory are added. That's out of xpcc's influence, since we haven't written our own libc (yet! ;-). But you can play with some of these settings and check `scons size` for impact.

> my own configuration file for this platform that seems to work, but I'm not sure where it fits in within XPCC.

Custom openocd config files can be placed in [tools/openocd/xpcc](https://github.com/roboterclubaachen/xpcc/tree/develop/tools/openocd/xpcc) which are then [accessible via `xpcc/file.cfg` to all projects](https://github.com/roboterclubaachen/xpcc/blob/develop/src/xpcc/architecture/platform/board/stm32f103c8t6_blue_pill/board.cfg#L10). Not necessary anymore for this PR, so just FYI.

salkinium commented on src/xpcc/architecture/platform/board/nucleo_f031k6/nucleo_f031k6.hpp:
```diff
@@ -0,0 +1,125 @@
+	static constexpr uint32_t Timer3  = ApbTimer;
+	static constexpr uint32_t Timer4  = ApbTimer;
+	static constexpr uint32_t Timer5  = ApbTimer;
+	static constexpr uint32_t Timer6  = ApbTimer;
```
I think F031 has TIM1, TIM2, TIM3, TIM14, TIM16, TIM17 all on APB. Otherwise :+1:.

salkinium commented on src/xpcc/architecture/platform/board/nucleo_f031k6/nucleo_f031k6.hpp:
```diff
@@ -0,0 +1,125 @@
+	enable()
+	{
+		ClockControl::enableInternalClock();	// 8MHz
+		// internal clock / 2 * 16 = 64MHz, => 64/1.5 = 42.6 => bad for USB
```
Update comments

salkinium commented on src/xpcc/architecture/platform/board/nucleo_f031k6/nucleo_f031k6.hpp:
```diff
@@ -0,0 +1,125 @@
+//
+// NUCLEO-F103RB
+// Nucleo kit for STM32F103RB
+// http://www.st.com/web/catalog/tools/FM116/SC959/SS1532/LN1847/PF259875
```
Update comments

salkinium commented on src/xpcc/architecture/platform/board/nucleo_f031k6/nucleo_f031k6.hpp:
```diff
@@ -0,0 +1,125 @@
+/* Copyright (c) 2016, Roboterclub Aachen e.V.
```
Put your name here and also 2017.

salkinium commented on src/xpcc/architecture/platform/board/nucleo_f031k6/nucleo_f031k6.hpp:
```diff
@@ -0,0 +1,125 @@
+using D12 = GpioB4;
+using D13 = GpioB3;
+
+using Button = xpcc::GpioInverted<GpioInputA12>;
```
I think the Nucleo-32 only has a Reset button, use `xpcc::GpioUnused` here instead.

genbattle replied:
The board I got from STM had a shunt across D2/GND, acting in the same capacity as the user button on the bigger boards. I wasn't aware of xpcc::GpioUnused though, so I will use that instead.

salkinium commented on src/xpcc/architecture/platform/board/nucleo_f031k6/nucleo_f031k6.hpp:
```diff
@@ -0,0 +1,125 @@
+using A4 = GpioA5;
+using A5 = GpioA6;
+using A6 = GpioA7;
+using A7 = GpioA2;
```
A6, A7 not in pinout.

genbattle replied:
Ah, I added these in because they were on the Nucleo-32 Pinout. I'll remove them.

salkinium replied:
Sorry, I'm full of crap, I misread the PDF, these pins are in it and the F303K8 one is wrong too. 🙄

genbattle commented:
> I'm impressed by your contribution considering how under-documented most of this is ;-P

The CONTRIBUTING.md document was helpful in getting started in this direction, but there was indeed quite a bit of wading through the code to get to this point. I did see the DFG, but didn't know where to source the device definition XML files from.

> would you be ok with licensing your contributions as MPLv2 as well?

Yes, that would be fine with me.

genbattle commented:
I've updated my changes in response to your review feedback, please let me know if there's anything I missed.

I wasn't quite sure what the correct way to update the copyright notice was, so I updated the date and added my name in addition to the existing one.

salkinium commented on src/xpcc/architecture/platform/board/nucleo_f031k6/board.cfg:
```diff
@@ -0,0 +1,16 @@
+device = stm32f031k6
+
+[parameters]
+uart.stm32.2.tx_buffer = 2048
```
256B

salkinium commented:
> I wasn't quite sure what the correct way to update the copyright notice was, so I updated the date and added my name in addition to the existing one.

I'd prefer just your name without the Roboterclub. We were lazy in xpcc when assigning copyright just to "Roboterclub Aachen e.V.", but in German law one can never relinquish copyright to anyone (but it can be permissively licensed of course). We fixed this in modm and added all original authors (thanks to git).

genbattle commented:
I've made the requested changes.

salkinium commented:
When you create the example, remember to add it to `.travis.yml` at the bottom, so that it's part of the CI.

genbattle commented:
I added A6 and A7 back into the board definition, I forgot to do that last night.

genbattle commented:
Added a blink example for the nucleo f031k6.

salkinium commented:
Don't forget `.travis.yml`

genbattle commented:
Apologies, I didn't see that comment earlier. I have added the example to travis.yml.

salkinium commented:
Ok, this looks good to me, I'll add all remaining device files and then rebase your PR to work again.

salkinium commented:
Squashed and rebased the commits.

salkinium commented:
Wow. This thing really doesn't have a lot of RAM…
```
Size after:
Memory Usage
------------
Device: stm32f031k6

Program:    3212 bytes (9.8% used)
(.data + .fastcode + .fastdata + .reset + .rodata + .text)

Data:       1528 bytes (37.3% used) = 504 bytes static (12.3%) + 1024 bytes stack (25.0%)
(.bss + .data + .fastcode + .fastdata + .noinit + .stack + .vectors)

Heap:       2628 bytes (64.2% available)
(.heap1)
```

genbattle commented:
It was even worse before the changes to reduce memory usage 😅

On 20 Mar 2017 7:27 AM, "Niklas Hauser" <notifications@github.com> wrote:

> Merged #221 <https://github.com/roboterclubaachen/xpcc/pull/221>.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/roboterclubaachen/xpcc/pull/221#event-1005939747>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAdgL8tz64j2xWjSH8vAF6L2vXle6_4eks5rnXOsgaJpZM4MZR-6>
> .
>
